File: cpp/115.distinct-subsequences.cpp
/*
 * @lc app=leetcode id=115 lang=cpp
 *
 * [115] Distinct Subsequences
 */
#include <string>
#include <unordered_map>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numDistinct(string s, string t)
    {
        if (s == t)
        {
            return 1;
        }
        if (!check_has_substring(s, t))
        {
            return 0;
        }
        int res = 0;
        unordered_map<char, vector<string>> pre_word_map;
        unordered_map<string, long> match_word_count;
        int len = t.size();
        for (int i = 1; i != len; ++i)
        {
            auto pre_word = t.substr(0, i);
            pre_word_map[t[i]].push_back(pre_word);
        }
        auto first_word_string = t.substr(0, 1);
        for (auto c : s)
        {
            for (auto iter = pre_word_map[c].rbegin(), iter_end = pre_word_map[c].rend(); iter != iter_end; ++iter)
            {
                auto pre_word = *iter;
                auto extended_pre_word = pre_word + c;
                match_word_count[extended_pre_word] += match_word_count[pre_word];
            }
            if (c == t[0])
            {
                ++match_word_count[first_word_string];
            }
        }
        return match_word_count[t];
    }
    bool check_has_substring(string &s, string &t)
    {
        if (t.size() > s.size())
        {
            return false;
        }
        int pos = 0;
        for (auto c : t)
        {
            auto next_pos = s.find(c, pos);
            if (next_pos == -1)
            {
                return false;
            }
            else
            {
                pos = next_pos + 1;
            }
        }
        return true;
    }
};
// @lc code=end

File: java/115.distinct-subsequences.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=115 lang=java
 *
 * [115] Distinct Subsequences
 */

// @lc code=start
class Solution {
    public int numDistinct(String s, String t) {
        if (s.equals(t)) {
            return 1;
        }
        if (!check_has_substring(s, t)) {
            return 0;
        }
        Map<Character, ArrayList<String>> pre_word_map = new HashMap<>();
        Map<String, Long> match_word_count = new HashMap<>();
        int t_len = t.length();
        for (int i = 1; i < t_len; ++i) {
            var pre_word = t.substring(0, i);
            if (!pre_word_map.containsKey(t.charAt(i))) {
                pre_word_map.put(t.charAt(i), new ArrayList<>());
            }
            pre_word_map.get(t.charAt(i)).add(pre_word);
        }
        var first_word_string = t.substring(0, 1);
        int s_len = s.length();
        for (int i = 0; i < s_len; ++i) {
            int candidate_size = pre_word_map.getOrDefault(s.charAt(i), new ArrayList<>()).size();
            for (int j = candidate_size - 1; j >= 0; --j) {
                var pre_word = pre_word_map.get(s.charAt(i)).get(j);
                var extended_pre_word = pre_word + s.charAt(i);
                match_word_count.put(extended_pre_word, match_word_count.getOrDefault(extended_pre_word, 0L)
                        + match_word_count.getOrDefault(pre_word, 0L));
            }
            if (s.charAt(i) == t.charAt(0)) {
                match_word_count.put(first_word_string, match_word_count.getOrDefault(first_word_string, 0L) + 1);
            }
        }
        return match_word_count.get(t).intValue();
    }

    boolean check_has_substring(String s, String t) {
        if (t.length() > s.length()) {
            return false;
        }
        int pos = 0;
        int t_len = t.length();
        for (int i = 0; i != t_len; ++i) {
            var next_pos = s.indexOf(t.charAt(i), pos);
            if (next_pos == -1) {
                return false;
            } else {
                pos = next_pos + 1;
            }
        }
        return true;
    }
}
// @lc code=end
